src/user_crate/mod.rs

```diff
@@ -210,12 +211,52 @@ fn parse_source_and_deps(code_and_deps: &str) -> eyre::Result<(syn::Block, toml:
     let user_dependencies: toml::value::Table =
         toml::from_str(&deps_block).map_err(PlRustError::ParsingDependenciesBlock)?;
 
+    if crate::gucs::allow_listed_dependencies_only() {
+        validate_dependences_are_allowed(&user_dependencies);
+    }
+
     let user_code: syn::Block =
         syn::parse_str(&code_block).map_err(PlRustError::ParsingCodeBlock)?;
 
     Ok((user_code, user_dependencies))
 }
 
+#[tracing::instrument(level = "debug", skip_all)]
+fn validate_dependences_are_allowed(user_dependencies: &toml::value::Table) {
+    let allowed_deps = crate::gucs::get_allow_listed_dependencies();
+    let mut unsupported_deps = std::collections::HashMap::<String, toml::value::Value>::new();
+
+    for (dep, version) in user_dependencies {
+        match version {
+            toml::Value::String(ver) => {
+                let req = semver::VersionReq::parse(ver.as_str()).unwrap();
+
+                if allowed_deps.contains_key(dep)
+                    && req.matches(
+                        &semver::Version::parse(&allowed_deps.get(dep).unwrap().as_str().unwrap())
+                            .unwrap(),
+                    )
+                {
+                    // version is supported
+                } else {
+                    unsupported_deps.insert(dep.to_string(), version.clone());
+                }
+            }
+            _ => {
+                pgx::error!("The toml::Value {:?} is not currently supported", version);
```

Comment:
We should be able to handle the "inline table" format and touch the "version" key inside it, as this is commonly used by Rust programmers. You can see an example of a similar block of code here: https://github.com/tcdi/pgx/blob/99f92951f630b51e0c32f00e8db3602b43f961d0/pgx-version-updater/src/main.rs#L262-L299

This way we are prepared for e.g. when we want to qualify which features of a crate are allowed.

Reply:
Added a basic table support for "version", not familiar with what values could potentially be used in an inline table, I'm guessing it can be arbitrary, in which case we probably need to add something recursive eventually

Reply:
Ahh thanks. I was wondering if something like this was possible:

```
pgx = {version = "0.6.0", features = ["a", "b", "c"], random_crate_specific_key = ["e", "f"]}
```

But looking at the manifest it doesn't seem like it.

Reply:
It's possible there's a key that has a fairly "wildcard" set of entries (`features` is basically like this) but afaik all the valid keys at this level are determined ahead of time.